File: 0695-max-area-of-island/0695-max-area-of-island.java
class Solution {
    public int max;
    public int count;
    public boolean arr[][];
    
    public int maxAreaOfIsland(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        arr = new boolean[m][n];
        for (int i = 0;i<m;i++)
        {
            for (int j = 0;j<n;j++)
            {
                if (grid[i][j] == 1)
                {
                    count = 0;
                    doSomething(grid, m, n, i, j);
                    max = Math.max(max, count);
                }
            }
        }
        return max;
    }
    
    public void doSomething(int[][] grid, int m, int n, int i, int j)
    {
        if (i < 0 || i == m || j < 0 || j == n)
        {
            return;
        }
        if (grid[i][j] == 0)
        {
            return;
        }
        if (arr[i][j] == true)
        {
            return;
        }
        arr[i][j] = true;
        count++;
        doSomething(grid, m, n, i, j+1);
        doSomething(grid, m, n, i+1, j);
        doSomething(grid, m, n, i, j-1);
        doSomething(grid, m, n, i-1, j);
        grid[i][j] = 0;
        return;
    }
}

File: 0695-max-area-of-island/0695-max-area-of-island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        maxarea = 0
        def dp(i, j):
            if i < 0 or i >= len(grid) or j < 0 or j >= len(grid[0]) or grid[i][j] == 0:
                return 0
            grid[i][j] = 0
            return 1 + dp(i, j + 1) + dp(i + 1, j) + dp(i, j - 1) + dp(i - 1, j)

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1:
                    temparea = dp(i, j)
                    maxarea = max(maxarea, temparea)
        return maxarea